File: clients/ts/src/protocol.ts
/**
 * PowDB wire protocol.
 *
 * Frame format: [type(1)][flags(1)][len(4 LE)][payload]
 * Strings are encoded as [len(4 LE)][utf-8 bytes].
 *
 * This mirrors crates/server/src/protocol.rs — keep them in sync.
 */

export const MSG_CONNECT = 0x01;
export const MSG_CONNECT_OK = 0x02;
export const MSG_QUERY = 0x03;
export const MSG_RESULT_ROWS = 0x07;
export const MSG_RESULT_SCALAR = 0x08;
export const MSG_RESULT_OK = 0x09;
export const MSG_ERROR = 0x0a;
export const MSG_DISCONNECT = 0x10;

export type Message =
  | { type: "Connect"; dbName: string; password: string | null }
  | { type: "ConnectOk"; version: string }
  | { type: "Query"; query: string }
  | { type: "ResultRows"; columns: string[]; rows: string[][] }
  | { type: "ResultScalar"; value: string }
  | { type: "ResultOk"; affected: bigint }
  | { type: "Error"; message: string }
  | { type: "Disconnect" };

// ───── Encoding ────────────────────────────────────────────────────────────

export function encode(msg: Message): Buffer {
  let msgType: number;
  let payload: Buffer;

  switch (msg.type) {
    case "Connect": {
      const dbBuf = encodeString(msg.dbName);
      const pwBuf =
        msg.password === null ? u32LE(0) : encodeString(msg.password);
      payload = Buffer.concat([dbBuf, pwBuf]);
      msgType = MSG_CONNECT;
      break;
    }
    case "ConnectOk":
      payload = encodeString(msg.version);
      msgType = MSG_CONNECT_OK;
      break;
    case "Query":
      payload = encodeString(msg.query);
      msgType = MSG_QUERY;
      break;
    case "ResultRows": {
      const parts: Buffer[] = [];
      const colCount = Buffer.alloc(2);
      colCount.writeUInt16LE(msg.columns.length, 0);
      parts.push(colCount);
      for (const col of msg.columns) parts.push(encodeString(col));
      parts.push(u32LE(msg.rows.length));
      for (const row of msg.rows) {
        for (const val of row) parts.push(encodeString(val));
      }
      payload = Buffer.concat(parts);
      msgType = MSG_RESULT_ROWS;
      break;
    }
    case "ResultScalar":
      payload = encodeString(msg.value);
      msgType = MSG_RESULT_SCALAR;
      break;
    case "ResultOk": {
      payload = Buffer.alloc(8);
      payload.writeBigUInt64LE(msg.affected, 0);
      msgType = MSG_RESULT_OK;
      break;
    }
    case "Error":
      payload = encodeString(msg.message);
      msgType = MSG_ERROR;
      break;
    case "Disconnect":
      payload = Buffer.alloc(0);
      msgType = MSG_DISCONNECT;
      break;
  }

  const frame = Buffer.alloc(6 + payload.length);
  frame.writeUInt8(msgType, 0);
  frame.writeUInt8(0, 1); // flags
  frame.writeUInt32LE(payload.length, 2);
  payload.copy(frame, 6);
  return frame;
}

// ───── Decoding ────────────────────────────────────────────────────────────

/**
 * Attempts to parse a single frame from the start of `buf`. Returns the parsed
 * message and the number of bytes consumed, or `null` if the buffer does not
 * yet contain a complete frame.
 */
export function tryDecode(
  buf: Buffer,
): { msg: Message; consumed: number } | null {
  if (buf.length < 6) return null;
  const msgType = buf.readUInt8(0);
  // flags byte at offset 1 is currently unused
  const payloadLen = buf.readUInt32LE(2);
  if (buf.length < 6 + payloadLen) return null;
  const payload = buf.subarray(6, 6 + payloadLen);
  const msg = decodePayload(msgType, payload);
  return { msg, consumed: 6 + payloadLen };
}

function decodePayload(msgType: number, payload: Buffer): Message {
  const cursor = { pos: 0 };
  switch (msgType) {
    case MSG_CONNECT: {
      const dbName = decodeString(payload, cursor);
      let password: string | null = null;
      if (cursor.pos < payload.length) {
        const p = decodeString(payload, cursor);
        password = p.length === 0 ? null : p;
      }
      return { type: "Connect", dbName, password };
    }
    case MSG_CONNECT_OK:
      return { type: "ConnectOk", version: decodeString(payload, cursor) };
    case MSG_QUERY:
      return { type: "Query", query: decodeString(payload, cursor) };
    case MSG_RESULT_ROWS: {
      const colCount = payload.readUInt16LE(cursor.pos);
      cursor.pos += 2;
      const columns: string[] = [];
      for (let i = 0; i < colCount; i++) {
        columns.push(decodeString(payload, cursor));
      }
      const rowCount = payload.readUInt32LE(cursor.pos);
      cursor.pos += 4;
      const rows: string[][] = [];
      for (let r = 0; r < rowCount; r++) {
        const row: string[] = [];
        for (let c = 0; c < colCount; c++) {
          row.push(decodeString(payload, cursor));
        }
        rows.push(row);
      }
      return { type: "ResultRows", columns, rows };
    }
    case MSG_RESULT_SCALAR:
      return { type: "ResultScalar", value: decodeString(payload, cursor) };
    case MSG_RESULT_OK: {
      const affected = payload.readBigUInt64LE(0);
      return { type: "ResultOk", affected };
    }
    case MSG_ERROR:
      return { type: "Error", message: decodeString(payload, cursor) };
    case MSG_DISCONNECT:
      return { type: "Disconnect" };
    default:
      throw new Error(`unknown message type: 0x${msgType.toString(16)}`);
  }
}

// ───── String helpers ──────────────────────────────────────────────────────

function encodeString(s: string): Buffer {
  const bytes = Buffer.from(s, "utf8");
  const out = Buffer.alloc(4 + bytes.length);
  out.writeUInt32LE(bytes.length, 0);
  bytes.copy(out, 4);
  return out;
}

function decodeString(buf: Buffer, cursor: { pos: number }): string {
  if (cursor.pos + 4 > buf.length) {
    throw new Error("truncated string length");
  }
  const len = buf.readUInt32LE(cursor.pos);
  cursor.pos += 4;
  if (cursor.pos + len > buf.length) {
    throw new Error("truncated string data");
  }
  const s = buf.toString("utf8", cursor.pos, cursor.pos + len);
  cursor.pos += len;
  return s;
}

function u32LE(n: number): Buffer {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(n, 0);
  return b;
}

File: clients/go/protocol.go
// Package powdb implements a Go client for PowDB.
//
// Frame format on the wire: [type(1)][flags(1)][len(4 LE)][payload].
// Strings are encoded as [len(4 LE)][utf-8 bytes].
//
// Mirrors crates/server/src/protocol.rs.
package powdb

import (
	"encoding/binary"
	"errors"
	"fmt"
)

const (
	msgConnect      byte = 0x01
	msgConnectOk    byte = 0x02
	msgQuery        byte = 0x03
	msgResultRows   byte = 0x07
	msgResultScalar byte = 0x08
	msgResultOk     byte = 0x09
	msgError        byte = 0x0A
	msgDisconnect   byte = 0x10
)

// Message is the union of all wire frames.
type Message interface{ isMessage() }

// Connect opens a session.
type Connect struct {
	DBName   string
	Password *string
}

// ConnectOk is the server's handshake reply.
type ConnectOk struct{ Version string }

// Query carries a PowQL statement.
type Query struct{ Query string }

// ResultRows is a tabular result.
type ResultRows struct {
	Columns []string
	Rows    [][]string
}

// ResultScalar is a single-value result.
type ResultScalar struct{ Value string }

// ResultOk indicates a write with the number of affected rows.
type ResultOk struct{ Affected uint64 }

// Error is a server-side error reply.
type Error struct{ Message string }

// Disconnect ends the session.
type Disconnect struct{}

func (Connect) isMessage()      {}
func (ConnectOk) isMessage()    {}
func (Query) isMessage()        {}
func (ResultRows) isMessage()   {}
func (ResultScalar) isMessage() {}
func (ResultOk) isMessage()     {}
func (Error) isMessage()        {}
func (Disconnect) isMessage()   {}

// Encode serialises a message into a full frame.
func Encode(m Message) []byte {
	var msgType byte
	var payload []byte

	switch v := m.(type) {
	case Connect:
		payload = encodeString(v.DBName)
		if v.Password == nil {
			payload = append(payload, 0, 0, 0, 0)
		} else {
			payload = append(payload, encodeString(*v.Password)...)
		}
		msgType = msgConnect
	case ConnectOk:
		payload = encodeString(v.Version)
		msgType = msgConnectOk
	case Query:
		payload = encodeString(v.Query)
		msgType = msgQuery
	case ResultRows:
		payload = make([]byte, 2)
		binary.LittleEndian.PutUint16(payload, uint16(len(v.Columns)))
		for _, c := range v.Columns {
			payload = append(payload, encodeString(c)...)
		}
		rc := make([]byte, 4)
		binary.LittleEndian.PutUint32(rc, uint32(len(v.Rows)))
		payload = append(payload, rc...)
		for _, row := range v.Rows {
			for _, val := range row {
				payload = append(payload, encodeString(val)...)
			}
		}
		msgType = msgResultRows
	case ResultScalar:
		payload = encodeString(v.Value)
		msgType = msgResultScalar
	case ResultOk:
		payload = make([]byte, 8)
		binary.LittleEndian.PutUint64(payload, v.Affected)
		msgType = msgResultOk
	case Error:
		payload = encodeString(v.Message)
		msgType = msgError
	case Disconnect:
		msgType = msgDisconnect
	default:
		panic(fmt.Sprintf("powdb: unknown message type %T", v))
	}

	frame := make([]byte, 6+len(payload))
	frame[0] = msgType
	frame[1] = 0 // flags
	binary.LittleEndian.PutUint32(frame[2:6], uint32(len(payload)))
	copy(frame[6:], payload)
	return frame
}

// TryDecode parses a single frame from the front of buf.
//
// If buf does not yet contain a complete frame, it returns (nil, 0, nil).
// On success it returns the message and the number of bytes consumed.
func TryDecode(buf []byte) (Message, int, error) {
	if len(buf) < 6 {
		return nil, 0, nil
	}
	msgType := buf[0]
	payloadLen := int(binary.LittleEndian.Uint32(buf[2:6]))
	if len(buf) < 6+payloadLen {
		return nil, 0, nil
	}
	payload := buf[6 : 6+payloadLen]
	msg, err := decodePayload(msgType, payload)
	if err != nil {
		return nil, 0, err
	}
	return msg, 6 + payloadLen, nil
}

func decodePayload(msgType byte, payload []byte) (Message, error) {
	switch msgType {
	case msgConnect:
		pos := 0
		db, err := decodeString(payload, &pos)
		if err != nil {
			return nil, err
		}
		var pw *string
		if pos < len(payload) {
			p, err := decodeString(payload, &pos)
			if err != nil {
				return nil, err
			}
			if p != "" {
				pw = &p
			}
		}
		return Connect{DBName: db, Password: pw}, nil
	case msgConnectOk:
		pos := 0
		v, err := decodeString(payload, &pos)
		if err != nil {
			return nil, err
		}
		return ConnectOk{Version: v}, nil
	case msgQuery:
		pos := 0
		q, err := decodeString(payload, &pos)
		if err != nil {
			return nil, err
		}
		return Query{Query: q}, nil
	case msgResultRows:
		if len(payload) < 2 {
			return nil, errors.New("truncated column count")
		}
		pos := 0
		colCount := int(binary.LittleEndian.Uint16(payload[pos:]))
		pos += 2
		cols := make([]string, colCount)
		for i := 0; i < colCount; i++ {
			s, err := decodeString(payload, &pos)
			if err != nil {
				return nil, err
			}
			cols[i] = s
		}
		if pos+4 > len(payload) {
			return nil, errors.New("truncated row count")
		}
		rowCount := int(binary.LittleEndian.Uint32(payload[pos:]))
		pos += 4
		rows := make([][]string, rowCount)
		for r := 0; r < rowCount; r++ {
			row := make([]string, colCount)
			for c := 0; c < colCount; c++ {
				s, err := decodeString(payload, &pos)
				if err != nil {
					return nil, err
				}
				row[c] = s
			}
			rows[r] = row
		}
		return ResultRows{Columns: cols, Rows: rows}, nil
	case msgResultScalar:
		pos := 0
		v, err := decodeString(payload, &pos)
		if err != nil {
			return nil, err
		}
		return ResultScalar{Value: v}, nil
	case msgResultOk:
		if len(payload) < 8 {
			return nil, errors.New("truncated result ok payload")
		}
		return ResultOk{Affected: binary.LittleEndian.Uint64(payload[:8])}, nil
	case msgError:
		pos := 0
		m, err := decodeString(payload, &pos)
		if err != nil {
			return nil, err
		}
		return Error{Message: m}, nil
	case msgDisconnect:
		return Disconnect{}, nil
	default:
		return nil, fmt.Errorf("unknown message type: 0x%x", msgType)
	}
}

func encodeString(s string) []byte {
	out := make([]byte, 4+len(s))
	binary.LittleEndian.PutUint32(out, uint32(len(s)))
	copy(out[4:], s)
	return out
}

func decodeString(buf []byte, pos *int) (string, error) {
	if *pos+4 > len(buf) {
		return "", errors.New("truncated string length")
	}
	l := int(binary.LittleEndian.Uint32(buf[*pos:]))
	*pos += 4
	if *pos+l > len(buf) {
		return "", errors.New("truncated string data")
	}
	s := string(buf[*pos : *pos+l])
	*pos += l
	return s, nil
}
